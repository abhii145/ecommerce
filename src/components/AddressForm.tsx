import React from "react"
import { useForm } from "react-hook-form"
import { z } from "zod"
import { zodResolver } from "@hookform/resolvers/zod"
import { Customer } from "../types"

interface AddressFormProps {
  onChange: (addressData: Customer) => void
  isSubmitting: boolean
  onBack: () => void
}

const customerSchema = z.object({
  name: z.string().min(1, "Name is required"),
  mobile: z
    .string()
    .regex(/^\d{10}$/, "Mobile number must be exactly 10 digits"),
  address: z.string().min(1, "Address is required"),
  email: z.string().email("Invalid email address"),
  pincode: z.string().regex(/^\d{6}$/, "Pincode must be exactly 6 digits"),
})

const AddressForm: React.FC<AddressFormProps> = ({
  onChange,
  isSubmitting,
  onBack,
}) => {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<Customer>({
    resolver: zodResolver(customerSchema),
  })

  const onSubmit = (data: Customer) => {
    onChange(data)
  }

  return (
    <div className="bg-white rounded-md px-4 py-6 h-max">
      <form className="space-y-4" onSubmit={handleSubmit(onSubmit)} noValidate>
        {(
          ["name", "mobile", "address", "email", "pincode"] as Array<
            keyof Customer
          >
        ).map((field) => (
          <div key={field} className="relative z-0 mt-4">
            <input
              id={field}
              type={field === "email" ? "email" : "text"}
              {...register(field)}
              placeholder=" "
              autoComplete="off"
              aria-invalid={errors[field] ? "true" : "false"}
              aria-describedby={`${field}-error`}
              className={`peer block py-2.5 px-0 w-full text-sm text-gray-900 bg-transparent border-0 border-b-2 border-gray-300 appearance-none focus:outline-none ${
                errors[field] ? "border-red-500" : ""
              }`}
            />
            <label
              htmlFor={field}
              className={`absolute text-sm text-gray-500 duration-300 transform -translate-y-6 scale-75 top-3 -z-10 origin-[0] peer-focus:left-0 peer-placeholder-shown:scale-100 peer-placeholder-shown:translate-y-0 peer-focus:scale-75 peer-focus:-translate-y-6 ${
                errors[field] ? "text-red-500" : ""
              }`}
              style={{ zIndex: 1 }}
            >
              {field.charAt(0).toUpperCase() + field.slice(1)}
            </label>
            {errors[field] && (
              <p id={`${field}-error`} className="text-red-500 text-sm">
                {errors[field]?.message}
              </p>
            )}
          </div>
        ))}
        <div className="flex space-x-2">
          <button
            type="button"
            className="text-sm px-2 py-2 w-1/3 font-semibold tracking-wide bg-blue-500 hover:bg-blue-600 text-white rounded-md"
            onClick={onBack}
            aria-label="Go back to the previous step"
          >
            Back
          </button>
          <button
            type="submit"
            className="text-sm px-4 py-2.5 w-2/3 font-semibold tracking-wide bg-gray-800 hover:bg-gray-900 text-white rounded-md"
            disabled={isSubmitting}
            aria-busy={isSubmitting}
            aria-label="Proceed to payment"
          >
            {isSubmitting ? "Loading..." : "Proceed to Payment"}
          </button>
        </div>
      </form>
    </div>
  )
}

export default AddressForm
