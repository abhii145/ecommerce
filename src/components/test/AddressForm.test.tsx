import React from "react"
import { render, screen, fireEvent, act } from "@testing-library/react"
import AddressForm from "../AddressForm"

describe("AddressForm Component", () => {
  const mockOnChange = jest.fn()
  const mockOnBack = jest.fn()

  beforeEach(() => {
    render(
      <AddressForm
        onChange={mockOnChange}
        isSubmitting={false}
        onBack={mockOnBack}
      />
    )
  })


  it("matches snapshot", () => {
    const { asFragment } = render(
      <AddressForm
        onChange={mockOnChange}
        isSubmitting={false}
        onBack={mockOnBack}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })


  it("should render all input fields", () => {
    const fields = ["Name", "Mobile", "Address", "Email", "Pincode"]
    fields.forEach((field) => {
      expect(screen.getByLabelText(field)).toBeInTheDocument()
    })
  })

  it("should update state on input change", () => {
    const nameInput = screen.getByLabelText("Name")
    fireEvent.change(nameInput, { target: { value: "John Doe" } })
    expect(nameInput).toHaveValue("John Doe")
  })

  it("should call onChange with correct data when all fields are filled", async () => {
    const nameInput = screen.getByLabelText("Name")
    const mobileInput = screen.getByLabelText("Mobile")
    const addressInput = screen.getByLabelText("Address")
    const emailInput = screen.getByLabelText("Email")
    const pincodeInput = screen.getByLabelText("Pincode")

    await act(async () => {
      fireEvent.change(nameInput, { target: { value: "John Doe" } })
      fireEvent.change(mobileInput, { target: { value: "1234567890" } })
      fireEvent.change(addressInput, { target: { value: "123 Main St" } })
      fireEvent.change(emailInput, { target: { value: "john@example.com" } })
      fireEvent.change(pincodeInput, { target: { value: "123456" } })
      fireEvent.submit(
        screen.getByRole("button", { name: /proceed to payment/i })
      )
    })

    expect(mockOnChange).toHaveBeenCalledWith({
      name: "John Doe",
      mobile: "1234567890",
      address: "123 Main St",
      pincode: "123456",
      email: "john@example.com",
    })
  })

  it("should call onBack when the back button is clicked", () => {
    fireEvent.click(
      screen.getByRole("button", { name: /go back to the previous step/i })
    )
    expect(mockOnBack).toHaveBeenCalled()
  })

  it("should show validation errors for empty fields", async () => {
    await act(async () => {
      fireEvent.submit(
        screen.getByRole("button", { name: /proceed to payment/i })
      )
    })

    expect(screen.getByText("Name is required")).toBeInTheDocument()
    expect(
      screen.getByText("Mobile number must be exactly 10 digits")
    ).toBeInTheDocument()
    expect(screen.getByText("Address is required")).toBeInTheDocument()
    expect(screen.getByText("Invalid email address")).toBeInTheDocument()
    expect(
      screen.getByText("Pincode must be exactly 6 digits")
    ).toBeInTheDocument()
  })

})
