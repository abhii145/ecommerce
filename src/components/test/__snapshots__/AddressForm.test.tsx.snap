// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressForm Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="bg-white rounded-md px-4 py-6 h-max"
  >
    <form
      class="space-y-4"
      novalidate=""
    >
      <div
        class="relative z-0 mt-4"
      >
        <input
          aria-describedby="name-error"
          aria-invalid="false"
          autocomplete="off"
          class="peer block py-2.5 px-0 w-full text-sm text-gray-900 bg-transparent border-0 border-b-2 border-gray-300 appearance-none focus:outline-none "
          id="name"
          name="name"
          placeholder=" "
          type="text"
        />
        <label
          class="absolute text-sm text-gray-500 duration-300 transform -translate-y-6 scale-75 top-3 -z-10 origin-[0] peer-focus:left-0 peer-placeholder-shown:scale-100 peer-placeholder-shown:translate-y-0 peer-focus:scale-75 peer-focus:-translate-y-6 "
          for="name"
          style="z-index: 1;"
        >
          Name
        </label>
      </div>
      <div
        class="relative z-0 mt-4"
      >
        <input
          aria-describedby="mobile-error"
          aria-invalid="false"
          autocomplete="off"
          class="peer block py-2.5 px-0 w-full text-sm text-gray-900 bg-transparent border-0 border-b-2 border-gray-300 appearance-none focus:outline-none "
          id="mobile"
          name="mobile"
          placeholder=" "
          type="text"
        />
        <label
          class="absolute text-sm text-gray-500 duration-300 transform -translate-y-6 scale-75 top-3 -z-10 origin-[0] peer-focus:left-0 peer-placeholder-shown:scale-100 peer-placeholder-shown:translate-y-0 peer-focus:scale-75 peer-focus:-translate-y-6 "
          for="mobile"
          style="z-index: 1;"
        >
          Mobile
        </label>
      </div>
      <div
        class="relative z-0 mt-4"
      >
        <input
          aria-describedby="address-error"
          aria-invalid="false"
          autocomplete="off"
          class="peer block py-2.5 px-0 w-full text-sm text-gray-900 bg-transparent border-0 border-b-2 border-gray-300 appearance-none focus:outline-none "
          id="address"
          name="address"
          placeholder=" "
          type="text"
        />
        <label
          class="absolute text-sm text-gray-500 duration-300 transform -translate-y-6 scale-75 top-3 -z-10 origin-[0] peer-focus:left-0 peer-placeholder-shown:scale-100 peer-placeholder-shown:translate-y-0 peer-focus:scale-75 peer-focus:-translate-y-6 "
          for="address"
          style="z-index: 1;"
        >
          Address
        </label>
      </div>
      <div
        class="relative z-0 mt-4"
      >
        <input
          aria-describedby="email-error"
          aria-invalid="false"
          autocomplete="off"
          class="peer block py-2.5 px-0 w-full text-sm text-gray-900 bg-transparent border-0 border-b-2 border-gray-300 appearance-none focus:outline-none "
          id="email"
          name="email"
          placeholder=" "
          type="email"
        />
        <label
          class="absolute text-sm text-gray-500 duration-300 transform -translate-y-6 scale-75 top-3 -z-10 origin-[0] peer-focus:left-0 peer-placeholder-shown:scale-100 peer-placeholder-shown:translate-y-0 peer-focus:scale-75 peer-focus:-translate-y-6 "
          for="email"
          style="z-index: 1;"
        >
          Email
        </label>
      </div>
      <div
        class="relative z-0 mt-4"
      >
        <input
          aria-describedby="pincode-error"
          aria-invalid="false"
          autocomplete="off"
          class="peer block py-2.5 px-0 w-full text-sm text-gray-900 bg-transparent border-0 border-b-2 border-gray-300 appearance-none focus:outline-none "
          id="pincode"
          name="pincode"
          placeholder=" "
          type="text"
        />
        <label
          class="absolute text-sm text-gray-500 duration-300 transform -translate-y-6 scale-75 top-3 -z-10 origin-[0] peer-focus:left-0 peer-placeholder-shown:scale-100 peer-placeholder-shown:translate-y-0 peer-focus:scale-75 peer-focus:-translate-y-6 "
          for="pincode"
          style="z-index: 1;"
        >
          Pincode
        </label>
      </div>
      <div
        class="flex space-x-2"
      >
        <button
          aria-label="Go back to the previous step"
          class="text-sm px-2 py-2 w-1/3 font-semibold tracking-wide bg-blue-500 hover:bg-blue-600 text-white rounded-md"
          type="button"
        >
          Back
        </button>
        <button
          aria-busy="false"
          aria-label="Proceed to payment"
          class="text-sm px-4 py-2.5 w-2/3 font-semibold tracking-wide bg-gray-800 hover:bg-gray-900 text-white rounded-md"
          type="submit"
        >
          Proceed to Payment
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
